mlir/lib/Dialect/MIOpen/Tuning/solver/conv_hip_implicit_gemm_bwd_v1r1.cpp

```diff
@@ -0,0 +1,555 @@
+    return std::make_tuple(-1, -1, -1, -1, failure());
+  }
+
+  if (!(GemmMPerBlock % GemmMPerThread == 0 &&
+        GemmNPerBlock % GemmNPerThread == 0))
+    return std::make_tuple(-1, -1, -1, -1, failure());
+
+  const auto thread_gemm_per_block_m = GemmMPerBlock / GemmMPerThread;
+  const auto thread_gemm_per_block_n = GemmNPerBlock / GemmNPerThread;
+
+  const auto thread_gemm_per_cluster_m =
+      GemmMLevel0Cluster * GemmMLevel1Cluster;
+  const auto thread_gemm_per_cluster_n =
+      GemmNLevel0Cluster * GemmNLevel1Cluster;
+
+  if (!(thread_gemm_per_block_m % thread_gemm_per_cluster_m == 0) &&
+      (thread_gemm_per_block_n % thread_gemm_per_cluster_n == 0))
+    return std::make_tuple(-1, -1, -1, -1, failure());
+
+  const auto cluster_per_block_m =
+      thread_gemm_per_block_m / thread_gemm_per_cluster_m;
+  const auto cluster_per_block_n =
+      thread_gemm_per_block_n / thread_gemm_per_cluster_n;
+
+  // inline asm only support cluster_per_block_m = 2 andcluster_per_block_n = 2
+  if (!(cluster_per_block_m == 2 && cluster_per_block_n == 2))
+    return std::make_tuple(-1, -1, -1, -1, failure());
+
+  return std::make_tuple(GemmMLevel0Cluster, GemmNLevel0Cluster,
+                         GemmMLevel1Cluster, GemmNLevel1Cluster, success());
+}
+
+std::tuple<int64_t, int64_t, int64_t, int64_t, LogicalResult>
+PerformanceImplicitGemmBwdDataV1R1::
+    CalculateGemmABlockCopyPerformanceParameters(
+        const ConvolutionContext &ctx) const {
+  int64_t ClusterLengths_GemmK = 0;
+  int64_t ClusterLengths_GemmM = 0;
+  int64_t SrcDataPerRead_GemmM = 4;
+  int64_t DstDataPerWrite_GemmM = 4;
+  int64_t DstDataPerWrite_GemmKPACK = 4;
+
+  // calculate vector length on gemmk dimension
+  SrcDataPerRead_GemmM =
+      ImplicitGemmUtil::gcd(SrcDataPerRead_GemmM, GemmMPerBlock);
+
+  // calculate threadwise copy size
+  const auto a_data_per_thread_copy =
+      (GemmKPerBlock * GemmMPerBlock) / BlockSize;
+
+  if (!(a_data_per_thread_copy > 0))
+    return std::make_tuple(-1, -1, -1, -1, failure());
+
+  // GemmABlockCopySrcDataPerRead_GemmK also bounded by size of threadwise copy
+  SrcDataPerRead_GemmM =
+      ImplicitGemmUtil::gcd(SrcDataPerRead_GemmM, a_data_per_thread_copy);
+
+  // decide threadwise copy lengths
+  const auto a_data_per_thread_copy_gemmm = SrcDataPerRead_GemmM;
+  const auto a_data_per_thread_copy_gemmk =
+      a_data_per_thread_copy / a_data_per_thread_copy_gemmm;
+
+  if (1) { // if (ctx.IsFp32())
+    DstDataPerWrite_GemmM = ImplicitGemmUtil::gcd(DstDataPerWrite_GemmM,
+                                                  a_data_per_thread_copy_gemmm);
+  }
+
+  // calculate blockwise copy thread cluster lengths
+  ClusterLengths_GemmK = GemmKPerBlock / a_data_per_thread_copy_gemmk;
+  ClusterLengths_GemmM = GemmMPerBlock / a_data_per_thread_copy_gemmm;
+
+  if (!(ClusterLengths_GemmK > 0 && ClusterLengths_GemmM > 0))
+    return std::make_tuple(-1, -1, -1, -1, failure());
+
+  if (1) { // if(ctx.IsFp32())
```

Comment:
In last scan-through, I find those `if (1)`s. Could you adopt data type into the convolution context if there isn't one already?

Reply:
It's best that you also add the data type check into `isValid` just so that the solver doesn't pick up wrong data types.

Reply:
@yiqian1 Will file an issue on bring up fp16 + fp32 tuning parameters.